File: leet-code/go/0278 first-bad-version.go
package leet

func isBadVersion(version int) bool

func firstBadVersion(n int) int {
	return binarySearch1(1, n)
}

func binarySearch1(start, end int) int {
	length := end - start
	pos := (length / 2) + start

	if length == 0 {
		return pos
	}

	if isBadVersion(pos) {
		return binarySearch1(start, pos)
	}

	return binarySearch1(pos+1, end)
}

File: leet-code/python/0278_first_bad_version.py
# The isBadVersion API is already defined for you.
# def isBadVersion(version: int) -> bool:

class Solution:
  def firstBadVersion(self, n: int) -> int:
    l, r = 1, n
    while l <= r:
      m = (l + r) // 2
      
      isBad = isBadVersion(m)
      if isBad:
        r = m - 1
      else:
        l = m + 1
    
    return l